src/lib/operators/join_nested_loop.cpp

```diff
@@ -0,0 +1,239 @@
+    auto nullable = (left_may_produce_null || _left_in_table->column_is_nullable(column_id));
+    _output_table->add_column_definition(_left_in_table->column_name(column_id), _left_in_table->column_type(column_id),
+                                         nullable);
+  }
+
+  // Preparing output table by adding columns from right table
+  for (ColumnID column_id{0}; column_id < _right_in_table->column_count(); ++column_id) {
+    auto nullable = (right_may_produce_null || _right_in_table->column_is_nullable(column_id));
+    _output_table->add_column_definition(_right_in_table->column_name(column_id),
+                                         _right_in_table->column_type(column_id), nullable);
+  }
+}
+
+template <typename LeftType, typename RightType>
+void JoinNestedLoop::_perform_join() {
+  auto left_table = _left_in_table;
+  auto right_table = _right_in_table;
+
+  auto left_column_id = _left_column_id;
+  auto right_column_id = _right_column_id;
+
+  auto _comparator = get_comparator<LeftType, RightType>(_scan_type);
+
+  if (_mode == JoinMode::Right) {
+    // for Right Outer we swap the tables so we have the outer on the "left"
+    left_table = _right_in_table;
+    right_table = _left_in_table;
+
+    left_column_id = _right_column_id;
+    right_column_id = _left_column_id;
+  }
+
+  auto pos_list_left = std::make_shared<PosList>();
+  auto pos_list_right = std::make_shared<PosList>();
+
+  const auto is_outer_join = (_mode == JoinMode::Left || _mode == JoinMode::Right || _mode == JoinMode::Outer);
+
+  // for Full Outer, remember the matches on the right side
+  std::set<RowID> right_matches;
+
+  // Scan all chunks from left input
+  for (ChunkID chunk_id_left = ChunkID{0}; chunk_id_left < left_table->chunk_count(); ++chunk_id_left) {
+    auto column_left = left_table->get_chunk(chunk_id_left).get_column(left_column_id);
+
+    resolve_column_type<LeftType>(*column_left, [&](const auto& typed_left_column) {
+      auto iterable_left = create_iterable_from_column<LeftType>(typed_left_column);
+
+      // for Outer joins, remember matches on the left side
+      std::vector<bool> left_matches;
+
+      if (is_outer_join) {
+        left_matches.resize(typed_left_column.size());
+      }
+
+      // Scan all chunks for right input
+      for (ChunkID chunk_id_right = ChunkID{0}; chunk_id_right < right_table->chunk_count(); ++chunk_id_right) {
+        auto column_right = right_table->get_chunk(chunk_id_right).get_column(right_column_id);
+
+        resolve_column_type<RightType>(*column_right, [&](const auto& typed_right_column) {
+          auto iterable_right = create_iterable_from_column<RightType>(typed_right_column);
+
+          iterable_left.for_each([&](const auto& left_value) {
+            if (left_value.is_null()) return;
+
+            iterable_right.for_each([&](const auto& right_value) {
+              if (right_value.is_null()) return;
+
+              if (_comparator(left_value.value(), right_value.value())) {
```

Comment:
If we want it still a bit faster, we could replace this `std::function` that resembles a virtual method call with something like `BaseTableScanImpl::_with_operator`.

Reply:
That's a good point, even though I am not sure how much performance that will actually get us, if any. And because I think that the current version is more readable, I would prefer to keep it as is (see Markus' comment above).

Reply:
What would the alternative be? (bear with me, I am still learning for loops)

If we really think the NLJ will be used in several query plans (as MD said), we should not that easily dismiss optimizations imo. But only if it's not 50 more lines of code or so.